extensions/amp-app-banner/1.0/test/test-amp-app-banner.js

```diff
@@ -0,0 +1,53 @@
+import {htmlFor} from '#core/dom/static-template';
+
+import {toggleExperiment} from '#experiments';
+
+import '../amp-app-banner';
+import {waitFor} from '#testing/helpers/service';
+
+describes.realWin(
+  'amp-app-banner-v1.0',
+  {
+    amp: {
+      extensions: ['amp-app-banner:1.0'],
+    },
+  },
+  (env) => {
+    let win;
+    let doc;
+    let html;
+
+    beforeEach(async () => {
+      win = env.win;
+      doc = win.document;
+      html = htmlFor(doc);
+      toggleExperiment(win, 'bento-app-banner', true, true);
+    });
+
+    async function mountElement(element) {
+      doc.body.appendChild(element);
+      await element.buildInternal();
+      await waitFor(() => element.isConnected, 'element connected');
+      return element;
+    }
+
+    it('should not render an element without an id', async () => {
+      const err = /bento-app-banner should have an id/;
+      expectAsyncConsoleError(err, 2);
+      await expect(
+        mountElement(html`<amp-app-banner></amp-app-banner>`)
+      ).to.be.rejectedWith(err);
+    });
+
+    it('renders the element', async () => {
+      await mountElement(html`
+        <amp-app-banner id="TEST">
+          <h2>Our app is way better</h2>
+          <button open-button>Get the app</button>
+        </amp-app-banner>
+      `);
+
+      // TODO: add assertions here
```

Comment:
So in theory, we should be able to use `env.sandbox.spy` (which is a wrapper for sinon) to spy on the `location.assign`. Then we can assert that it was called with the correct link to the app or play store.

This is all just theory however, because I haven't found an example of anyone using this to spy on `location.assign` before.

Reply:
I figured out how to improve these tests.  First I added mocks for `platformService`, then had to fix some logic for querying the light-dom from within the shadow-dom.
Long story short, I've got some tests here now, and it's working great!